doc/design/block.md

```diff
@@ -0,0 +1,128 @@
+# Block design
+
+In computer programming, a block is a lexical structure of source code which is grouped together.
+In most programming languages, block is useful when define a function or some conditional statements such as `if` and `while`.
+
+In Paddle, we need a similar concept to support following scenes:
+
+- the basic unit of execution, just like the original `NetOp`, block can group multiple operators and make them behave just like a single operator. Unlike `NetOp`, a network should not be an operator, but a block can be treated as a operator.
+  - Block should have a `Exec` method, which will execute the operators grouped in the block.
+- store the context for operators, including the information of variables they will operator on. 
+- Support the definition and execution of special operators such as `RNNOp`, `switch_op`, `while_op`, `if_else_op`, which needs one or more code-blocks.
+  - Block will help to define and execute thouse operators.
+  
+## when need a block
+Blocks are needed in following code snippets
+```c++
+if(cond) { block_true } else { block_false }
+```
+
+```c++
+while(cond){ block }
+```
+
+```c++
+int func(inputs) { block with outputs }
+```
+
+```c++
+namespace xxxx {
+  block
+}
+```
+
+Paddle will have `if_else_op`, `switch_op`, `while_op` and some other operators that need a function block such as RNNOp, and enable definition and execution of a block is vital to those operators.
+
+What's more, Paddle should has a `pd.namespace`, which will help to preventing name conflicts of variables in large model.
+
+For example:
+
+```python
+import paddle as pd
+
+def submodel(namespace, x):
+    # begin a namespace
+    with pd.namespace(namespace):
+        # should be complex network, here easy one for demo
+        # variable W's name is namespace + "W"
+        # so in different namespace, there is a W without conflicts
+        W = pd.get_variable("W", shape=[30, 30], reuse=True)
+        y = pd.matmul(W, x)
+        return y
+
+inputs = [a, b, c]
+
+# create 3 submodels and each one has its own parameters.
+a_out = submodel("model_a", a)
+b_out = submodel("model_b", b)
+c_out = submodel("model_c", c)
+
+
+# d is some op's output
+d = someop()
+
+# reuse model_a
+d_model_a_out = submodel("model_a", d)
+```
+
+with `pd.namespace`, user create 3 different namespaces and in each one, the parameters those have specific name will be reused.
+
+## Why use Block to replace NetOp
```

Comment:
What's the lifecycle of a block?

Reply:
currently, block is maintained by python, no harm I think

Reply:
I think the lifecycle of a Block is a very good question and a must-to-answer question in this design doc.  Because Block is a C++ class, its lifecycle should be discussed in the scope of the C++ core.